include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,1221 @@
+ * deletes the underlying comm ptr
+ */
+class comm_deleter {
+public:
+    using pointer = MPI_Comm*;
+    void operator()(pointer comm) const
+    {
+        GKO_ASSERT(*comm != MPI_COMM_NULL);
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_free(comm));
+        delete comm;
+    }
+};
+
+
+}  // namespace
+
+
+/**
+ * A communicator class that takes in the given communicator and duplicates it
+ * for our purposes. As the class or object goes out of scope, the communicator
+ * is freed.
+ */
+class communicator {
+public:
+    /**
+     * Non-owning constructor for an existing communicator of type MPI_Comm. The
+     * MPI_Comm object will not be deleted after the communicator object has
+     * been freed and an explicit MPI_Comm_free needs to be called on the
+     * original MPI_Comm object.
+     *
+     * @param comm The input MPI_Comm object.
+     */
+    communicator(const MPI_Comm& comm)
+    {
+        this->comm_.reset(new MPI_Comm(comm));
+        this->size_ = get_num_ranks();
+        this->rank_ = get_my_rank();
+        this->node_local_rank_ = get_node_local_rank();
+    }
+
+    /**
+     * Create a communicator object from an existing MPI_Comm object using color
+     * and key.
+     *
+     * @param comm The input MPI_Comm object.
+     * @param color The color to split the original comm object
+     * @param key  The key to split the comm object
+     */
+    communicator(const MPI_Comm& comm, int color, int key)
+    {
+        MPI_Comm comm_out;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_split(comm, color, key, &comm_out));
+        this->comm_.reset(new MPI_Comm(comm_out), comm_deleter{});
+        this->size_ = get_num_ranks();
+        this->rank_ = get_my_rank();
+        this->node_local_rank_ = get_node_local_rank();
+    }
+
+    /**
+     * Create a communicator object from an existing MPI_Comm object using color
+     * and key.
+     *
+     * @param comm The input communicator object.
+     * @param color The color to split the original comm object
+     * @param key  The key to split the comm object
+     */
+    communicator(const communicator& comm, int color, int key)
+    {
+        MPI_Comm comm_out;
+        GKO_ASSERT_NO_MPI_ERRORS(
+            MPI_Comm_split(comm.get(), color, key, &comm_out));
+        this->comm_.reset(new MPI_Comm(comm_out), comm_deleter{});
+        this->size_ = get_num_ranks();
+        this->rank_ = get_my_rank();
+        this->node_local_rank_ = get_node_local_rank();
+    }
+
+    communicator(const communicator& other) = default;
+
+    communicator& operator=(const communicator& other) = default;
+
+    communicator(communicator&& other)
+    {
+        this->comm_ = std::move(other.comm_);
+        other.comm_.reset(new MPI_Comm(MPI_COMM_NULL));
+    }
+
+    communicator& operator=(communicator&& other)
+    {
+        this->comm_ = std::move(other.comm_);
+        other.comm_.reset(new MPI_Comm(MPI_COMM_NULL));
+        return *this;
+    }
```

Comment:
use copy semantics here, there should be no empty `communicator` instances!
```suggestion
    communicator(communicator&& other) : communicator(other) {}

    communicator& operator=(communicator&& other)
    {
        return *this = other;
    }
```

Reply:
In that case, maybe it's better to not have move constructors and assignments at all? Just keep the copy constructor and copy assignment.

Reply:
yeah, I just wasn't sure whether this causes any issues when you do try to move, but it should be fine.

Reply:
yes, sorry, I was trying something out and forgot to remove the code.